tamr_toolbox/dataset/manage.py

```diff
@@ -0,0 +1,527 @@
+    The default attribute type will be ARRAY STRING. To set non-default attribute types, you must
+    define them in the attribute_types dictionary. Any attribute descriptions can be specified in
+    the attribute_descriptions dictionary. By default, the existing attribute types will not
+    change unless override_existing_types is set to True. When it's False, the attribute type
+    updates will only be logged.
+
+    Args:
+        dataset: An existing TUC dataset
+        attributes: Complete list of attribute names that should exist in the updated dataset
+        attribute_types: dictionary for non-default types, attribute name is the key and
+            AttributeType is the value
+        attribute_descriptions: dictionary for attribute descriptions, attribute name is the
+            key and the attribute description is the value
+        description: updated description of dataset, if None will not update the description
+        tags: updated tags for the dataset, if None will not update tags
+        override_existing_types: boolean flag, when true will alter existing attribute's types
+
+    Returns:
+        Updated Dataset
+
+    Raises:
+        requests.HTTPError: If any HTTP error is encountered
+        ValueError: If the dataset is not a source dataset
+        TypeError: If the attributes argument is not an Iterable
+
+    Example:
+        >>> import tamr_toolbox as tbox
+        >>> from tbox.models import attribute_type
+        >>> tamr_client = tbox.utils.client.create(**instance_connection_info)
+        >>> dataset = = tamr_client.datasets.by_name("my_dataset_name")
+        >>> tbox.dataset.manage.update(
+        >>>     client=tamr_client,
+        >>>     dataset=dataset,
+        >>>     attributes=["unique_id","name","address","total_sales"],
+        >>>     attribute_types={"total_sales":attribute_type.ARRAY(attribute_type.DOUBLE)},
+        >>>     override_existing_types = True,
+        >>> )
+    """
+    dataset_name = dataset.name
+    if dataset.upstream_datasets():
+        raise ValueError(f"{dataset_name} is not a source dataset")
+    primary_keys = dataset.spec().to_dict()["keyAttributeNames"]
+
+    # Check input type is correct
+    if attributes and not isinstance(attributes, Iterable):
+        raise TypeError("attributes arg must be an Iterable")
+
+    # Update description and tags
+    dataset_spec = dataset.spec()
+    if description:
+        dataset_spec = dataset_spec.with_description(description)
+        LOGGER.info(f"Updating description for {dataset_name}")
+    if tags:
+        dataset_spec = dataset_spec.with_tags(tags)
+        LOGGER.info(f"Updating tags for {dataset_name}")
+
+    dataset_spec.put()
+
+    if attributes:
+        # Get current dataset attributes
+        target_dataset_attributes = dataset.attributes
+        existing_attributes = []
+        for attr in target_dataset_attributes.stream():
```

Comment:
nit: no need to call `stream` when iterating. Also this can be more concise
```suggestion
        existing_attributes = [attr.name for attr in dataset.attributes]
```